### PR TITLE
[AsianCrush] fix extractor, add support for yuyutv and Midnight Pulp

### PR DESCRIPTION
**Pull request by TemporaryUsername12481**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The AsianCrush extractor was partially broken (didn't extract description properly).  This pull request fixes that, and also adds support for sister sites yuyutv and Midnight Pulp (closing #21281).

### CODE REVIEW

**Comment by ealgase** [comment]
@remitamine sorry to bother you, but could you please take a look at this?  (I think I've been understanding the code standards better now, so hopefully you won't have to leave as much feedback on this)

**Comment by ealgase** [comment]
I can do that for the main AsianCrushIE, but for the Playlist IE, it currently requires an additional site specific variable (the `_SITE_TITLE`).

**Comment by dstftw** [comment]
Nothing stops from rewriting it to use `re.sub`.

**Comment by ealgase** [comment]
I don't understand?  I'm referring to the last bit of this:
```python
        title = remove_end(
            self._html_search_regex(
                r'(?s)<h1\b[^>]\bid=["\']movieTitle[^>]+>(.+?)</h1>', webpage,
                'title', default=None) or self._og_search_title(
                webpage, default=None) or self._html_search_meta(
                'twitter:title', webpage, 'title',
                default=None) or self._search_regex(
                r'<title>([^<]+)</title>', webpage, 'title', fatal=False),
            ' | %s' % self._SITE_TITLE)
```
I don't see how `re.sub` would allow that last bit to work without knowing the title of the site.

**Comment by dstftw** [comment]
`re.sub(r'\s*\|\s*.+?$', '', title)`.

**Comment by ealgase** [comment]
OK, I've made the requested changes.

File: youtube_dl/extractor/asiancrush.py [hunk 1]
```diff
@@ -5,34 +5,64 @@
...
-    extract_attributes,
-    remove_end,
-)
+from ..utils import extract_attributes
 
 
 class AsianCrushIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?asiancrush\.com/video/(?:[^/]+/)?0+(?P<id>\d+)v\b'
+    IE_NAME = 'asiancrush'
+    _DOMAINS = r'(?:asiancrush\.com|yuyutv\.com|midnightpulp\.com)'
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Inline everything used only once.

File: youtube_dl/extractor/asiancrush.py [hunk 2]
```diff
@@ -63,23 +93,47 @@ def _real_extract(self, url):
                 r'/p(?:artner_id)?/(\d+)', player, 'partner id',
                 default='513551')
 
-        return self.url_result(
-            'kaltura:%s:%s' % (partner_id, kaltura_id),
-            ie=KalturaIE.ie_key(), video_id=kaltura_id,
-            video_title=title)
+        description = self._html_search_regex(r'<div class="description">(.+?)</div>', webpage, 'description', fatal=False, flags=re.DOTALL)
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Move flags into regex. Carry long lines.

File: youtube_dl/extractor/asiancrush.py [hunk 3]
```diff
@@ -89,22 +143,22 @@ def _real_extract(self, url):
...
+
+        title = re.sub(r'\s*\|\s*.+?$', '',
+                       self._html_search_regex(
+                           r'(?s)<h1\b[^>]\bid=["\']movieTitle[^>]+>(.+?)</h1>', webpage,
+                           'title', default=None) or self._og_search_title(
+                           webpage, default=None) or self._html_search_meta(
+                           'twitter:title', webpage, 'title',
+                           default=None) or self._search_regex(
+                           r'<title>([^<]+)</title>', webpage, 'title', fatal=False)
+                       )
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Revert indent.

File: youtube_dl/extractor/asiancrush.py [hunk 4]
```diff
@@ -63,23 +93,47 @@ def _real_extract(self, url):
...
+            'id': video_id,
+            'title': title,
+            'description': description,
+        }
 
 
 class AsianCrushPlaylistIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?asiancrush\.com/series/0+(?P<id>\d+)s\b'
-    _TEST = {
+    _VIDEO_IE = AsianCrushIE
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: No.

File: youtube_dl/extractor/asiancrush.py [hunk 5]
```diff
@@ -89,22 +143,22 @@ def _real_extract(self, url):
         entries = []
 
         for mobj in re.finditer(
-                r'<a[^>]+href=(["\'])(?P<url>%s.*?)\1[^>]*>' % AsianCrushIE._VALID_URL,
+                r'<a[^>]+href=(["\'])(?P<url>%s.*?)\1[^>]*>' % self._VIDEO_IE._VALID_URL,
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Revert.

File: youtube_dl/extractor/asiancrush.py (hunk 2 above)
Comment (TemporaryUsername12481) [reply youtube_dl/extractor/asiancrush.py]: I'm confused about "Move flags into regex", I can't find a difference between the way this is implemented in other extractors.

Comment (dstftw) [reply youtube_dl/extractor/asiancrush.py]: What difference are you even talking about?

File: youtube_dl/extractor/asiancrush.py [hunk 6]
```diff
@@ -5,34 +5,63 @@
...
-    extract_attributes,
-    remove_end,
-)
+from ..utils import extract_attributes
 
 
 class AsianCrushIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?asiancrush\.com/video/(?:[^/]+/)?0+(?P<id>\d+)v\b'
+    IE_NAME = 'asiancrush'
+    _VALID_URL = r'https?://(?:www\.)?(?P<host>(?:asiancrush\.com|yuyutv\.com|midnightpulp\.com))/video/(?:[^/]+/)?0+(?P<id>\d+)v\b'
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Move .com part outside the inner group.

File: youtube_dl/extractor/asiancrush.py [hunk 7]
```diff
@@ -96,15 +148,16 @@ def _real_extract(self, url):
                 entries.append(self.url_result(
                     mobj.group('url'), ie=AsianCrushIE.ie_key()))
 
-        title = remove_end(
+        title = re.sub(
```
Comment (dstftw) [inline youtube_dl/extractor/asiancrush.py]: Breaks on None title.

File: youtube_dl/extractor/asiancrush.py (hunk 2 above)
Comment (TemporaryUsername12481) [reply youtube_dl/extractor/asiancrush.py]: You said "Move flags into regex", I don't understand what you're asking me to do.

Comment (dstftw) [reply youtube_dl/extractor/asiancrush.py]: Do you aware what flags are at all?

**Comment by dstftw** [comment]
Does not work:
```
> py -3.7 .\youtube_dl\__main__.py https://www.yuyutv.com/video/013886v/the-act-of-killing/ -v --proxy 127.0.0.1:8118
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.yuyutv.com/video/013886v/the-act-of-killing/', '-v']
[debug] Encodings: locale cp1251, fs utf-8, out utf-8, pref cp1251
[debug] youtube-dl version 2019.07.14
[debug] Git HEAD: 1ef4607
[debug] Python version 3.7.0 (CPython) - Windows-10-10.0.10240-SP0
[debug] exe versions: ffmpeg N-85653-gb4330a0, ffprobe N-85653-gb4330a0, phantomjs 2.1.1, rtmpdump 2.4
[asiancrush] 13886: Downloading webpage
[asiancrush] 13886: Downloading webpage
[Kaltura] 1_66x4rg7o: Downloading video info JSON
[Kaltura] 1_66x4rg7o: Downloading m3u8 information
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on 'http://cdnapi.kaltura.com/p/513551/sp/51355100/playManifest/entryId/1_66x4rg7o/format/url/protocol/http/flavorId/1_hgns56wd'
ERROR: unable to download video data: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1915, in process_info
    success = dl(filename, info_dict)
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1854, in dl
    return fd.download(name, info)
  File "C:\Dev\youtube-dl\master\youtube_dl\downloader\common.py", line 366, in download
    return self.real_download(filename, info_dict)
  File "C:\Dev\youtube-dl\master\youtube_dl\downloader\http.py", line 341, in real_download
    establish_connection()
  File "C:\Dev\youtube-dl\master\youtube_dl\downloader\http.py", line 109, in establish_connection
    ctx.data = self.ydl.urlopen(request)
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 2227, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "C:\Python\Python37\lib\urllib\request.py", line 531, in open
    response = meth(req, response)
  File "C:\Python\Python37\lib\urllib\request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\Python\Python37\lib\urllib\request.py", line 563, in error
    result = self._call_chain(*args)
  File "C:\Python\Python37\lib\urllib\request.py", line 503, in _call_chain
    result = func(*args)
  File "C:\Python\Python37\lib\urllib\request.py", line 755, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "C:\Python\Python37\lib\urllib\request.py", line 531, in open
    response = meth(req, response)
  File "C:\Python\Python37\lib\urllib\request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\Python\Python37\lib\urllib\request.py", line 569, in error
    return self._call_chain(*args)
  File "C:\Python\Python37\lib\urllib\request.py", line 503, in _call_chain
    result = func(*args)
  File "C:\Python\Python37\lib\urllib\request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

**Pull request closed without merging**